perma-payments/perma_payments/views.py

```diff
@@ -0,0 +1,428 @@
+
+
+@csrf_exempt
+@require_http_methods(["POST"])
+def subscription(request):
+    """
+    Returns a simplified version of a registrar's subscription status,
+    as needed for making decisions in Perma.
+    """
+    try:
+        data = process_perma_transmission(request.POST, FIELDS_REQUIRED_FROM_PERMA['subscription'])
+    except InvalidTransmissionException:
+        return bad_request(request)
+
+    standing_subscription = SubscriptionAgreement.registrar_standing_subscription(data['registrar'])
+    if not standing_subscription:
+        subscription = None
+    else:
+        subscription = {
+            'rate': standing_subscription.subscription_request.recurring_amount,
+            'frequency': standing_subscription.subscription_request.recurring_frequency
+        }
+
+        if standing_subscription.cancellation_requested:
+            subscription['status'] = 'Cancellation Requested'
+        else:
+            subscription['status'] = standing_subscription.status
+
+    response = {
+        'registrar': data['registrar'],
+        'subscription': subscription,
+        'timestamp': datetime.utcnow().timestamp()
+    }
+    return JsonResponse({'encrypted_data': prep_for_perma(response).decode('ascii')})
+
+
+@csrf_exempt
+@require_http_methods(["POST"])
+def cancel_request(request):
+    """
+    Records a cancellation request from Perma.cc
+    """
+    try:
+        data = process_perma_transmission(request.POST, FIELDS_REQUIRED_FROM_PERMA['cancel_request'])
+    except InvalidTransmissionException:
+        return bad_request(request)
+
+    registrar = data['registrar']
+
+    # The user must have a subscription that can be cancelled.
+    sa = SubscriptionAgreement.registrar_standing_subscription(registrar)
+    if not sa or not sa.can_be_altered():
+        return render(request, 'generic.html', {'heading': "We're Having Trouble With Your Cancellation Request",
+                                                'message': "We can't find any active subscriptions associated with your account.<br>" +
+                                                           "If you believe this is an error, please contact us at <a href='mailto:{0}?subject=Our%20Subscription'>{0}</a>.".format(settings.DEFAULT_CONTACT_EMAIL)})
+
+    context = {
+        'registrar': registrar,
+        'search_url': CS_SUBSCRIPTION_SEARCH_URL[settings.CS_MODE],
+        'perma_url': settings.PERMA_URL,
+        'registrar_detail_path': settings.REGISTRAR_DETAIL_PATH,
+        'registrar_users_path': settings.REGISTRAR_USERS_PATH,
+        'merchant_reference_number': sa.subscription_request.reference_number
+    }
+    logger.info("Cancellation request received from registrar {} for {}".format(registrar, context['merchant_reference_number']))
+    send_admin_email('ACTION REQUIRED: cancellation request received', settings.DEFAULT_FROM_EMAIL, request, template="email/cancel.txt", context=context)
```

Comment:
I forget, is this async? If it's async then the error may not propagate correctly

Reply:
Sorry, I don't follow. Which error?

Reply:
FWIW, it looks like Django sends emails synchronously.